src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -2500,7 +2501,9 @@ public void run() {
                     // nothing else we can safely do here
                 }
             } finally {
-                JBossExecutors.setContextClassLoader(currentThread, old);
+                if (old != ctxCL) {
+                    JBossExecutors.setContextClassLoader(currentThread, old);
+                }
```

Comment:
I suppose there's some risk here that something within the task has set the ContextClassLoader to an entirely new value (e.g. `executor.execute(() -> Thread.currentThread().setContextClassLoad(otherClassLoader))`, which will be leaked into the threads state.
Perhaps `setContextClassLoader` should check the current value before setting a new one instead of doing the check in `getAndSetContextClassLoader`?

It would be entirely reasonable for my use-cases to add a new static option to avoid context classloader checks entirely, but I don't know if that would be generally acceptable.

Reply:
> Perhaps setContextClassLoader should check the current value before setting a new one instead of doing the check in getAndSetContextClassLoader?

Good point; Yep, that was indeed my concern re changing the original behaviour i.e. it expect on task completion to restore whatever there was before, regardless any previous change.

I think believe performing 2 loads + checks to be better then 1 load + 2 stores

re

> It would be entirely reasonable for my use-cases to add a new static option to avoid context classloader checks entirely, but I don't know if that would be generally acceptable.

I don't know if the quarkus use case would fall into this, really @dmlloyd might know?